File: 78subset.js
var subsets = function (nums) {
  let ans = [];
  var rec = (i, arr) => {
    ans.push(arr);
    for (var j = i + 1; j < nums.length; j++) {
      rec(j, [nums[j], ...arr]);
    }
  };
  rec(-1, []);
  return ans;
};

console.log(subsets([1, 2, 3]));

File: 78subset.py
class Solution(object):
    def subsets(self, nums):
        self.res = []
        self.recursion(nums, 0, [])
        return self.res

    def recursion(self, nums, i, arr):
        if i >= len(nums):
            self.res.append(arr)
            return
        tmp1 = arr.copy()
        self.recursion(nums, i+1, tmp1)
        tmp2 = arr.copy()
        tmp2.append(nums[i])
        self.recursion(nums, i+1, tmp2)


a = Solution()
print(a.subsets([1, 2, 3]))
